File: Editorial/1055_Shortest_Way_to_Form_String/CPP/7_Bottom_Up_Dynamic_Programming_TLE.cpp
class Solution {
public:
    int shortestWay(string source, string target) {
        
        // Boolean array to mark all characters of source
        bool sourceChars[26] = {false};
        for (auto c : source) {
            sourceChars[c - 'a'] = true;
        }

        // Check if all characters of target are present in source
        // If any character is not present, return -1
        for (auto c : target) {
            if (!sourceChars[c - 'a']) {
                return -1;
            }
        }

        // Optimal Answer for a given ending index. Memoizing using an Array
        vector <int> memo(target.length(), INT_MAX);
        memo[0] = 1;

        for (int endingIndex = 1; endingIndex < target.length(); endingIndex++) {
            if (isSubsequence(0, endingIndex, source, target)) {
                memo[endingIndex] = 1;
            } else {
                for (int partitionIndex = endingIndex - 1; partitionIndex >= 0; partitionIndex--) {
                    if (memo[partitionIndex] != INT_MAX &&
                            isSubsequence(partitionIndex + 1, endingIndex, source, target)) {
                        memo[endingIndex] = min(memo[endingIndex], memo[partitionIndex] + 1);
                    }
                }
            }
        }

        return memo[target.length() - 1];
    }

    
    // For toCheck passing indices of target, both included.
    bool isSubsequence(int start, int end, string source, string target) {

        if end - start + 1 > source.length() {
            return false;
        }

        int i = start;
        int j = 0;

        while (i <= end && j < source.length()) {
            if (target[i] == source[j]) {
                i++;
            }
            j++;
        }

        return i == end + 1;
    }
};

File: Editorial/1001-1100/1055_Shortest_Way_to_Form_String/Python3/7_Bottom_Up_Dynamic_Programming_TLE.py
class Solution:
    def shortestWay(self, source: str, target: str) -> int:

        # Check if the Task is possible or not
        # Set of all characters of source. Can use boolean array too.
        source_chars = set(source)

        # Check if all characters of target are present in source
        # If any character is not present, return -1
        for char in target:
            if char not in source_chars:
                return -1

        # Modified is_subsequence function. in_string will always be source
        # For to_check, passing indices of target, both included
        def is_subsequence(start, end):

            i = start
            j = 0

            while i <= end and j < len(source):
                if target[i] == source[j]:
                    i += 1
                j += 1

            return i == end + 1

        # Optimal Answer Array for each ending Index
        opt = [float('inf')] * len(target)
        opt[0] = 1

        for ending_index in range(1, len(target)):
            if is_subsequence(0, ending_index):
                opt[ending_index] = 1
            else:
                for partition_index in range(ending_index):
                    if (opt[partition_index] != float('inf') and
                            is_subsequence(partition_index + 1, ending_index)):
                        opt[ending_index] = min(
                            opt[ending_index], opt[partition_index] + 1)

        # Return Optimal Answer for last index
        return opt[-1]
